### PR TITLE
[KEYCLOAK-17399] - Declarative User Profile and UI

### PR DESCRIPTION
**Pull request by pedroigor**

### CODE REVIEW

**Comment by pedroigor** [comment]
> This client technology specific property should not be here in generic/reusable Validation SPI. I believe client code can infer response status code from `validatorId` plus `message` if necessary.
> I expect you introduced this to return `401` from admin REST API when user tries to modify readOnly user attribute (an dit iis even not used in final code). As you have specific validator for this condition, you can easily infer REST response status code from `validatorId` in error, no need for additional field here.

I don't think that is true. Validators should be able to indicate the status code for their messages. As a hint. It is up to the client code to consider it or ignore it. Note that client code does not issue errors but receive them from validators, so they should respect the semantics associated with messages and defined by validators. Otherwise, you are going to create boilerplate code and couple clients to validators.

Also, it should be possible to issue different errors from the same validator.

As added in the comment, IMO messages should have additional metadata about how they should be handled. I think that is the optimal approach. But it would require more changes.

**Comment by pedroigor** [comment]
> Is this really used? Looks like `verify-profile.ftl` takes attributes metadata from `VerifyProfileBean`, not the one prepared here. So maybe this should be removed to have cleaner code?

Makes sense. Going to remove it.

**Comment by velias** [comment]
> > This client technology specific property should not be here in generic/reusable Validation SPI. I believe client code can infer response status code from `validatorId` plus `message` if necessary.
> > I expect you introduced this to return `401` from admin REST API when user tries to modify readOnly user attribute (an dit iis even not used in final code). As you have specific validator for this condition, you can easily infer REST response status code from `validatorId` in error, no need for additional field here.
> 
> I don't think that is true. Validators should be able to indicate the status code for their messages. As a hint. It is up to the client code to consider it or ignore it. Note that client code does not issue errors but receive them from validators, so they should respect the semantics associated with messages and defined by validators. Otherwise, you are going to create boilerplate code and couple clients to validators.
> 
> Also, it should be possible to issue different errors from the same validator.
> 
> As added in the comment, IMO messages should have additional metadata about how they should be handled. I think that is the optimal approach. But it would require more changes.

I don't agree ;-) Really universal validator can't decide how its error is interpreted, only calling code can do this. Imagine validator validating value against list of allowed values. How can it decide if input is bad value only entered by user into some form, or permission violation because calling code uses it for some permissions check? Only calling code knows what exactly it validates and how to interpret the result. 
And also universal validator may not be used in the context of REST API call only, so using JAX-RS class here is not fully correct if we want to keep validator SPI and validators universal. So either some independent statuses or bind it to messages as you propose.

But I agree this is not critical for now, can be updated later if necessary.

**Comment by pedroigor** [comment]
> > > This client technology specific property should not be here in generic/reusable Validation SPI. I believe client code can infer response status code from `validatorId` plus `message` if necessary.
> > > I expect you introduced this to return `401` from admin REST API when user tries to modify readOnly user attribute (an dit iis even not used in final code). As you have specific validator for this condition, you can easily infer REST response status code from `validatorId` in error, no need for additional field here.
> > 
> > 
> > I don't think that is true. Validators should be able to indicate the status code for their messages. As a hint. It is up to the client code to consider it or ignore it. Note that client code does not issue errors but receive them from validators, so they should respect the semantics associated with messages and defined by validators. Otherwise, you are going to create boilerplate code and couple clients to validators.
> > Also, it should be possible to issue different errors from the same validator.
> > As added in the comment, IMO messages should have additional metadata about how they should be handled. I think that is the optimal approach. But it would require more changes.
> 
> I don't agree ;-) Really universal validator can't decide how its error is interpreted, only calling code can do this. Imagine validator validating value against list of allowed values. How can it decide if input is bad value only entered by user into some form, or permission violation because calling code uses it for some permissions check? Only calling code knows what exactly it validates and how to interpret the result.
> And also universal validator may not be used in the context of REST API call only, so using JAX-RS class here is not fully correct if we want to keep validator SPI and validators universal. So either some independent statuses or bind it to messages as you propose.
> 
> But I agree this is not critical for now, can be updated later if necessary.

I see your point. As you noticed, having that as metadata in messages is probably the best approach. The validator is still universal and the status code is only a hint. The client code can still decide whether or not it wants to respect the hint from messages/validators.

However, I'm not sure if client code should decide the status code. This is really associated with the message (and right now we are doing that within the validator) and in order to keep consistency, messages should be handled in the same manner by any client code. For HTTP-based clients, they should just respect the hint. For internal validation, you can just ignore for now.

Let us review this later and come up with a better definition for messages and the metadata associated with them.

File: services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java [hunk 1]
```diff
@@ -227,53 +236,77 @@ protected UserProfileMetadata decorateUserProfileForCache(UserProfileMetadata me
...
                 // do not take requirements from config for username and email as they are
                 // driven by business logic from parent!
-
-                if (rc.isAlways() || UPConfigUtils.isRoleForContext(context, rc.getRoles())) {
+                required = AttributeMetadata.ALWAYS_TRUE;
+                if (rc.isAlways()) {
                     validators.add(createRequiredValidator(attrConfig));
-                    required = AttributeMetadata.ALWAYS_TRUE;
+                } else if (rc.getRoles() != null) {
+                    validators.add(new AttributeValidatorMetadata(AttributeRequiredByMetadataValidator.ID,
```
Comment (velias) [inline services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java:243]: `required` predicate is not set correctly here, so metadata used in forms are incorrect, they always mark attribute as required

File: services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java [hunk 2]
```diff
@@ -227,53 +236,77 @@ protected UserProfileMetadata decorateUserProfileForCache(UserProfileMetadata me
...
+                            ValidatorConfig.builder()
+                                    .config("roles", rc.getRoles())
+                                    .build()));
                 } else if (UPConfigUtils.canBeAuthFlowContext(context) && rc.getScopes() != null && !rc.getScopes().isEmpty()) {
                     // for contexts executed from auth flow and with configured scopes requirement
                     // we have to create required validation with scopes based selector
-                    required = (c) -> attributePredicateAuthFlowRequestedScope(rc.getScopes());
-                    validators.add(createRequiredValidator(attrConfig));
+                    validators.add(new AttributeValidatorMetadata(AttributeRequiredByMetadataValidator.ID,
```
Comment (velias) [inline services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java:250]: `required` predicate is not set correctly here, so metadata used in forms are incorrect, they always mark attribute as required

File: services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java (hunk 1 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileProvider.java]: Missed during the refactoring. Should be OK now.

Will add a test too.

**Comment by thomasdarimont** [comment]
I just gave this a quick spin. Unfortunately I already hit a first bug: 

When I try to create a new user profile via the admin-console, I see an error in the dev tools console.
image

**Comment by bs-matil** [comment]
@thomasdarimont  same here

**Comment by pedroigor** [comment]
@thomasdarimont I also see some errors in console, but they are not impacting the functionality for me. I'm going to remove those entries in the log.

W.r.t to the JSON Editor. I could not reproduce. I tried to copy and past a different config as well as remove all attributes and add a new config.

Could you give me more details how to reproduce the problem? Could that be related to browser/version? I have some issues in the test suite due to the usage of some specific ECMAScript constructs ...

**Comment by thomasdarimont** [comment]
@pedroigor I just updated my chrome to Version 91.0.4472.77 (Official Build) (64-bit) this morning.

I simply started Keycloak with the `declarative_user_profile` enabled and tried to create a new custom user profile definition.
```
$ cat standalone/configuration/profile.properties
# Both are needed for token exchange
feature.token_exchange=enabled
feature.admin_fine_grained_authz=enabled

feature.account_api=enabled

#feature.scripts=enabled
#feature.upload_scripts=enabled

feature.impersonation=enabled
feature.declarative_user_profile=enabled
```

I already see errors when I click on the "user profile" menu of the master realm in the admin-console.
image

**Comment by pedroigor** [comment]
@thomasdarimont I see now. It works when running in Undertow ... But it is failing now in Wildfly dist. Not sure why. Going to fix it.

**Comment by pedroigor** [comment]
@thomasdarimont Sorry ... It is a configuration issue. Please add this to `standalone.xml`:

```
<spi name="userProfile">
    <default-provider>declarative-user-profile</default-provider>
</spi>
```

Right now we have two providers: legacy and declarative. Legacy is always enabled by default, so you should set declarative one as the default.

In the future this should go away as we are going to have a single provider.

**Comment by pedroigor** [comment]
@thomasdarimont @bs-matil I'm working on docs but here are some things to consider when using this:

* We have a `VerifyProfile` required action. This action must be enabled to the realm. It is always triggered when the user is not in compliance with the user profile configuration.
* Other required actions and templates, and account console, are not yet updated to access user profile metadata. This is planned to the next version.

**Comment by bs-matil** [comment]
I can confirm that the basic setup works with your suggested fix. I will have a look into the details.

Regards the "other required actions and templates". Does this include things like the registration action?

**Comment by pedroigor** [comment]
> Regards the "other required actions and templates". Does this include things like the registration action?

Yes.

**Comment by bs-matil** [comment]
Yes I already hat a brief look into the new bean for the profile view, which I expect to end up in all other flows too.

**Comment by pedroigor** [comment]
> Yes I already hat a brief look into the new bean for the profile view, which I expect to end up in all other flows too.

There are some improvements to config itself for dynamic forms. For instance:

* `Display Name` for attributes
* `Ordering` for attributes
* `Hints`

We have been discussing to solve this either by annotations or having top-level constructs in the config. Let's see ...

**Comment by bs-matil** [comment]
> 
> 
> > Yes I already hat a brief look into the new bean for the profile view, which I expect to end up in all other flows too.
> 
> There are some improvements to config itself for dynamic forms. For instance:
> 
>     * `Display Name` for attributes
> 
>     * `Ordering` for attributes
> 
>     * `Hints`
> 
> 
> We have been discussing to solve this either by annotations or having top-level constructs in the config. Let's see ...

Last time this had been discussed the idea was to keep this loose coupled and keep it bound to the UI not the profile. So it would be annotations then. But I also see a lot value in some basic properties like order and the translation key which could be made explicit.

**Comment by thomasdarimont** [comment]
Basic setup is working for me too, thanks @pedroigor for the great work!

**Comment by thomasdarimont** [comment]
I just tried add a custom validation for emails, e.g. requiring that emails must end with `@local`
image

I noticed that the error message is not translated and has no field context. It only says `Bad Request: error-pattern-no-match` without mentioning the field.
image

**Comment by pedroigor** [comment]
> Basic setup is working for me too, thanks @pedroigor for the great work!

Our work. Including @velias.

**Comment by pedroigor** [comment]
As per validation, we might be missing some messages in the properties. Let me review what are they.

**Comment by pedroigor** [comment]
@thomasdarimont One thing though we need to discuss later is related to metadata associated with messages. For instance, HTTP status code.

Have a message is not enough and we should be considering binding metadata to messages so that client code can decide how to properly handle them. Right now, we changed the error in validation to pass a hint about the status code.

**Comment by pedroigor** [comment]
Documentation https://github.com/keycloak/keycloak-documentation/pull/1200.

**Review by miquelsi — APPROVED** [review]
LGTM. Thank you Pedro!

File: server-spi-private/src/main/java/org/keycloak/events/EventType.java [hunk 3]
```diff
@@ -63,6 +63,8 @@
     UPDATE_TOTP_ERROR(true),
     VERIFY_EMAIL(true),
     VERIFY_EMAIL_ERROR(true),
+    VERIFY_PROFILE(true),
+    VERIFY_PROFILE_ERROR(true),
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/events/EventType.java:67]: This is a bit weird - events should be create/update user, not verify user

Comment (pedroigor) [reply server-spi-private/src/main/java/org/keycloak/events/EventType.java:67]: See https://issues.redhat.com/browse/KEYCLOAK-18429.

**Comment by ghost81** [comment]
> <spi name="userProfile">
>     <default-provider>declarative-user-profile</default-provider>
> </spi>

How this could be achieved for keycloak.x distribution (quarkus based), as the same problem is persistent there too.

**Pull request merged**